### PR TITLE
openai[minor]: adds withStructuredOutput to chat oai

### PR DESCRIPTION
**Pull request by bracesproul**
Depends on https://github.com/langchain-ai/langchainjs/pull/4525 (merge & release core)

### CODE REVIEW

File: docs/core_docs/docs/integrations/chat/openai.mdx [hunk 1]
```diff
@@ -48,6 +48,30 @@ import OpenAITools from "@examples/models/chat/integration_openai_tool_calls.ts"
...
 
+### `.withStructuredOutput({ ... })`
+
+You can also use the `.withStructuredOutput({ ... })` method to coerce `ChatOpenAI` into returning a structured output.
+
+The method allows for passing in either a valid JSON schema (like what is returned from [`zodToJsonSchema`](https://www.npmjs.com/package/zod-to-json-schema)).
+
+Using the method is simple, just define your LLM and call `.withStructuredOutput({ ... })` with the desired schema.
+
+Here is an example using a ZOD schema and the `functionCalling` mode (default mode):
```
Comment (jacoblee93) [inline docs/core_docs/docs/integrations/chat/openai.mdx:59]: Why is ZOD all caps?

File: libs/langchain-openai/src/chat_models.ts [hunk 2]
```diff
@@ -843,4 +854,118 @@ export class ChatOpenAI<
...
+   * @template {z.ZodObject<any, any, any, any>} RunOutput The output type for the Runnable, expected to be a Zod schema object for structured output validation.
+   * @template {RunnableConfig} CallOptions The type for call options, extending from RunnableConfig.
+   *
+   * @param {z.ZodEffects<RunOutput>} schema The schema for the structured output. Either as a ZOD schema or a valid JSON schema object.
+   * @param {string} name The name of the function to call.
+   * @param {"functionCalling" | "jsonMode"} method The method to use for getting the structured output. Defaults to "functionCalling".
+   * @param {boolean | undefined} includeRaw Whether to include the raw output in the result. Defaults to false.
+   * @returns {Runnable<RunInput, RunOutput, CallOptions>} A new runnable that calls the LLM with structured output.
+   */
+  withStructuredOutput<
```
Comment (jacoblee93) [inline libs/langchain-openai/src/chat_models.ts:871]: Are we not adding this to the core interface? Down not to for now, just checking

File: libs/langchain-openai/src/chat_models.ts [hunk 3]
```diff
@@ -843,4 +854,118 @@ export class ChatOpenAI<
...
+
+    if (method === "jsonMode") {
+      llm = this.bind({
+        response_format: { type: "json_object" },
+      } as Partial<CallOptions>);
+      outputParser = new JsonOutputParser<RunOutput>();
+    } else {
+      // Is function calling
+      if (isZodSchema(schema)) {
+        class TmpClass extends StructuredTool {
```
Comment (jacoblee93) [inline libs/langchain-openai/src/chat_models.ts:900]: Can we just define this as JSON schema directly? This is hacky

File: libs/langchain-openai/src/chat_models.ts [hunk 4]
```diff
@@ -843,4 +854,118 @@ export class ChatOpenAI<
...
+          name = name ?? "";
+
+          async _call(input: RunInput) {
+            return JSON.stringify(input);
+          }
+        }
+        llm = this.bind({
+          tools: [new TmpClass()],
+          tool_choice: "auto",
+        } as unknown as Partial<CallOptions>);
```
Comment (jacoblee93) [inline libs/langchain-openai/src/chat_models.ts:916]: Why is this cast necessary?

File: libs/langchain-openai/src/chat_models.ts (hunk 3 above)
Comment (bracesproul) [reply libs/langchain-openai/src/chat_models.ts:900]: ya, just requires adding zodToJsonSchema as a hard dep which I wasn't sure we wanted to do.

File: libs/langchain-openai/src/chat_models.ts [hunk 5]
```diff
@@ -843,4 +854,118 @@ export class ChatOpenAI<
...
+    method?: "functionCalling" | "jsonMode";
+    includeRaw?: boolean;
+  }): Runnable<RunInput, RunOutput> {
+    let llm: Runnable;
+    let outputParser: JsonOutputKeyToolsParser | JsonOutputParser<RunOutput>;
+
+    if (method === "jsonMode") {
+      llm = this.bind({
+        response_format: { type: "json_object" },
+      } as Partial<CallOptions>);
```
Comment (jacoblee93) [inline libs/langchain-openai/src/chat_models.ts:895]: Are our types messed up?

File: libs/langchain-openai/src/chat_models.ts [hunk 6]
```diff
@@ -843,4 +854,118 @@ export class ChatOpenAI<
...
+            {
+              type: "function" as const,
+              function: {
+                name,
+                parameters: schema,
+              },
+            },
+          ],
+          tool_choice: "auto",
+        } as unknown as Partial<CallOptions>);
```
Comment (jacoblee93) [inline libs/langchain-openai/src/chat_models.ts:934]: We should fix our types, this is not good if people are running into this in the wild

File: libs/langchain-openai/src/chat_models.ts (hunk 4 above)
Comment (bracesproul) [reply libs/langchain-openai/src/chat_models.ts:916]: Was throwing errors without, as if it was going off the type of `.bind` on the base runnable class.

File: libs/langchain-openai/src/chat_models.ts [hunk 7]
```diff
@@ -843,4 +854,118 @@ export class ChatOpenAI<
...
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      parsed: (input: any) => outputParser.invoke(input.raw),
+    });
+    const parserNone = RunnablePassthrough.assign({
+      parsed: () => null,
+    });
+    const parsedWithFallback = parserAssign.withFallbacks({
+      fallbacks: [parserNone],
+    });
+    return new RunnableMap({
```
Comment (jacoblee93) [inline libs/langchain-openai/src/chat_models.ts:956]: Just use `RunnableSequence.from()` - not really intended to use the map constructor directly as it's a bit cumbersome

File: libs/langchain-openai/src/chat_models.ts [hunk 8]
```diff
@@ -843,4 +854,118 @@ export class ChatOpenAI<
...
+      }
+    }
+
+    if (!includeRaw) {
+      return llm.pipe(outputParser);
+    }
+
+    const parserAssign = RunnablePassthrough.assign({
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      parsed: (input: any) => outputParser.invoke(input.raw),
```
Comment (jacoblee93) [inline libs/langchain-openai/src/chat_models.ts:948]: Pretty sure this is not going to trace properly

File: libs/langchain-openai/src/chat_models.ts (hunk 3 above)
Comment (jacoblee93) [reply libs/langchain-openai/src/chat_models.ts:900]: I think fine to

File: libs/langchain-openai/src/chat_models.ts (hunk 4 above)
Comment (jacoblee93) [reply libs/langchain-openai/src/chat_models.ts:916]: Can we type it more strongly then? `let llm: Runnable<BaseLanguageModelInput, output, CallOptions>;`?

File: libs/langchain-openai/src/chat_models.ts (hunk 3 above)
Comment (bracesproul) [reply libs/langchain-openai/src/chat_models.ts:900]: nvm it already is!

File: libs/langchain-openai/src/chat_models.ts (hunk 2 above)
Comment (bracesproul) [reply libs/langchain-openai/src/chat_models.ts:871]: We are, but not all of these args, py only had schema, i had to add name for JS

File: docs/core_docs/docs/integrations/chat/openai.mdx (hunk 1 above)
Comment (bracesproul) [reply docs/core_docs/docs/integrations/chat/openai.mdx:59]: good catch, idk why I did that

File: libs/langchain-openai/src/chat_models.ts (hunk 8 above)
Comment (bracesproul) [reply libs/langchain-openai/src/chat_models.ts]: Fixed

File: docs/core_docs/docs/integrations/chat/openai.mdx (hunk 1 above)
Comment (jacoblee93) [reply docs/core_docs/docs/integrations/chat/openai.mdx:59]: It's still capitalized

File: libs/langchain-openai/src/chat_models.ts [hunk 9]
```diff
@@ -843,4 +852,118 @@ export class ChatOpenAI<
...
+   *
+   * @param {z.ZodEffects<RunOutput>} schema The schema for the structured output. Either as a Zod schema or a valid JSON schema object.
+   * @param {string} name The name of the function to call.
+   * @param {"functionCalling" | "jsonMode"} method The method to use for getting the structured output. Defaults to "functionCalling".
+   * @param {boolean | undefined} includeRaw Whether to include the raw output in the result. Defaults to false.
+   * @returns {Runnable<RunInput, RunOutput, CallOptions>} A new runnable that calls the LLM with structured output.
+   */
+  withStructuredOutput<
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    RunInput = any,
```
Comment (jacoblee93) [inline libs/langchain-openai/src/chat_models.ts:871]: You don't need this

File: libs/langchain-openai/src/chat_models.ts [hunk 10]
```diff
@@ -843,4 +852,118 @@ export class ChatOpenAI<
...
+    name,
+    method = "functionCalling",
+    includeRaw = false,
+  }: {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    schema: z.ZodEffects<RunOutput> | Record<string, any>;
+    name: string;
+    method?: "functionCalling" | "jsonMode";
+    includeRaw?: boolean;
+  }): Runnable<RunInput, RunOutput> {
```
Comment (jacoblee93) [inline libs/langchain-openai/src/chat_models.ts:886]: This output type is wrong? Should be:

```js
{
  parsed: RunOutput;
  raw: string;
}
```

File: examples/src/models/chat/integration_openai_wsa_json_schema.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  schema: calculatorSchema,
+  name: "calculator",
+  includeRaw: true,
+  method: "jsonMode",
+});
+
+const includeRawChain = prompt.pipe(includeRawModel);
+const includeRawResult = await includeRawChain.invoke({});
+console.log(JSON.stringify(includeRawResult, null, 2));
+/**
```
Comment (jacoblee93) [inline examples/src/models/chat/integration_openai_wsa_json_schema.ts:59]: Single star, it's not a docstring

**Pull request merged**